In [1]:
import os    
from dotenv  import load_dotenv
from pathlib import Path
import requests
import pandas as pd
pd.options.display.max_colwidth = 600  # Set Full Display

# Connect to API:

## First Key:

In [2]:
load_dotenv(dotenv_path = Path("C:/Users/Leon/API_keys/.env"))
auproperty_api=os.getenv('auproperty_vincent2')
type(auproperty_api)

str

In [3]:
auth_url = 'https://api2.aupropertyreport.com/api/auth'

auth_response = requests.post(auth_url, data = {
                        'email':"vincentgao076@gmail.com",
                        'api_key':auproperty_api,
                        })
auth_token = auth_response.json() 
access_token = auth_token['data']["token"]
auth = {"Authorization":"Bearer " + access_token}
auth

{'Authorization': 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL2FwaTIuYXVwcm9wZXJ0eXJlcG9ydC5jb20vYXBpL2F1dGgiLCJpYXQiOjE2NDg3MzE3NTEsImV4cCI6MTY0ODgxODE1MSwibmJmIjoxNjQ4NzMxNzUxLCJqdGkiOiJ4V0d0dGZsbWJJdmhtejZBIiwic3ViIjozNywicHJ2IjoiNDBhNDZkZjhmNTlkN2E5ZGVkZDc0NjRkZDI2NTNjOGIxNzE2ZTBiZCIsInBsYW4iOm51bGx9.rxVeYciPqrK0vkB4_NrhHs55FyygBkiMKNftanVhEeA'}

## Second Key:

In [4]:
# Overwrite, initiate second API key directly !!
# property_search_respond={}

In [5]:
property_search_url = "https://api2.aupropertyreport.com/api/property/search?q=3+17+hillcrest+ave+chadstone+vic+3148"

# GET Request
property_search_request = requests.get(property_search_url,headers=auth,params={"limit":"50"})
property_search_respond = property_search_request.json()
property_search_respond

{'message': 'You have exceeded your rate limit.', 'status_code': 429}

Exceed: {'message': 'You have exceeded your rate limit.', 'status_code': 429}

Normal:  [{'address': '3/17 Hillcrest Avenue, Chadstone Vic 3148',
  'code': 'unit-3-17-hillcrest-ave-chadstone-vic-3148',
  'state': 'VIC',
  'suburb': 'Chadstone',
  'postcode': '3148'}]

In [6]:
# 由于 address code search 的 respond 都是 list, 即使是空的，或者多个的，用来check rate-limit
if isinstance(property_search_respond, dict):

    auproperty_api=os.getenv('auproperty_vincent2')

    auth_url = 'https://api2.aupropertyreport.com/api/auth'

    auth_response = requests.post(auth_url, data = {
                        'email':"leon20130301@hotmail.com",
                        'api_key':os.getenv('auproperty_vincent'),
                        })
    auth_token = auth_response.json() 
    access_token = auth_token['data']["token"]
    auth = {"Authorization":"Bearer " + access_token}
    auth # 如果没有任何显示，则说明没有使用这个 API key
    print(auth) # 如果没有任何显示，则说明没有使用这个 API key

{'Authorization': 'Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL2FwaTIuYXVwcm9wZXJ0eXJlcG9ydC5jb20vYXBpL2F1dGgiLCJpYXQiOjE2NDg3MzE3NTQsImV4cCI6MTY0ODgxODE1NCwibmJmIjoxNjQ4NzMxNzU0LCJqdGkiOiJkMExUZ1JNM2FmQm5VeTZ0Iiwic3ViIjozNiwicHJ2IjoiNDBhNDZkZjhmNTlkN2E5ZGVkZDc0NjRkZDI2NTNjOGIxNzE2ZTBiZCIsInBsYW4iOiJtb250aGx5X3BsYW5fYV9zbyJ9.Gn_a4naWlUnC63iN3lPciWQzUXoqytzCGIMRd2sNNzU'}


# Import Specific Surburb Data:

'ASHWOOD', 'BALWYN', 'BALWYN NORTH', 'BLACKBURN',
'BLACKBURN SOUTH', 'BOX HILL', 'BOX HILL SOUTH', 'BURWOOD',
'BURWOOD EAST', 'CAMBERWELL', 'CHADSTONE', 'CLAYTON',
'CLAYTON SOUTH', 'DONCASTER', 'DONCASTER EAST', 'FOREST HILL',
'GLEN IRIS', 'GLEN WAVERLEY', 'HUNTINGDALE', 'MITCHAM',
'MONT ALBERT', 'MOUNT WAVERLEY', 'MULGRAVE', 'NOTTING HILL',
'NUNAWADING', 'OAKLEIGH', 'OAKLEIGH EAST', 'RINGWOOD',
'SURREY HILLS', 'TEMPLESTOWE', 'VERMONT', 'VERMONT SOUTH',
'WHEELERS HILL'

In [7]:
suburb="Glen_Waverley"

# use r -- to avoid unicode error !
file_path=Path(rf'C:\Users\LEON\test\domain\address of {suburb}.xlsx')
df = pd.read_excel(file_path, parse_dates=True)
print(df.info())
df.sample(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19479 entries, 0 to 19478
Data columns (total 8 columns):
flat_number         5238 non-null object
number_first        19479 non-null object
number_last         1700 non-null float64
street_name         19479 non-null object
street_type_code    18987 non-null object
locality_name       19479 non-null object
postcode            19479 non-null int64
address             19479 non-null object
dtypes: float64(1), int64(1), object(6)
memory usage: 1.2+ MB
None


flat_number number_first  number_last street_name street_type_code  \
18955         NaN           60          NaN     WATSONS             ROAD   

       locality_name  postcode                             address  
18955  GLEN WAVERLEY      3150  60 WATSONS ROAD GLEN WAVERLEY 3150

## Prepare address to the specific format for API call:

27A+Fermanagh+Road+Camberwell+vic+3124

In [8]:
df_target=df['address'].str.replace(" ","+").to_frame()
display(df_target.tail(2))

address
19477        6+CROWS+LANE+GLEN+WAVERLEY+3150
19478  11+MANNERING+DRIVE+GLEN+WAVERLEY+3150

# Get Address Code:

In [9]:
# TESTING:

# get index number of a specific row:
# df_target[df_target["address"]=='4+OBRIEN+CLOSE+GLEN+WAVERLEY+3150'].index

# select row by index:
# df_target.iloc[[0]]

# Convert to String:
# df_target.iloc[ind]["address"]

### Initialization: Loading saved data

In [10]:
# use r -- to avoid unicode error !
filesaved_path=Path(rf'C:\Users\LEON\test\domain\Address_Code_Output_{suburb}.xlsx')
filesaved_df = pd.read_excel(filesaved_path, parse_dates=True)
print(filesaved_df.info())
filesaved_df.tail(1)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1345 entries, 0 to 1344
Data columns (total 1 columns):
address_code    1345 non-null object
dtypes: object(1)
memory usage: 10.6+ KB
None


address_code
1344  4-swan-ct-glen-waverley-vic-3150

### Loop over dataframe and make API call:

In [11]:
'''
test2= pd.DataFrame([[{"a":1,"b":2}],[2],[3]])
test2
'''

'\ntest2= pd.DataFrame([[{"a":1,"b":2}],[2],[3]])\ntest2\n'

In [12]:
property_search_respond

{'message': 'You have exceeded your rate limit.', 'status_code': 429}

In [13]:
# Initialization:
list_address_code=[]
api_count=0

# Loop over dataframe and make API call:
for ind in df_target.index:
    if ind >= len(filesaved_df) and api_count <= 50-1: # 非常注意：从0开始，所以call 4个。
        key_word_addresscode = df_target.iloc[ind]["address"]
        property_search_url = f"https://api2.aupropertyreport.com/api/property/search?q={key_word_addresscode}"
        property_search_request = requests.get(property_search_url,headers=auth,params={"limit":"50"})
        property_search_respond = property_search_request.json()
        api_count += 1    
    
        if property_search_respond != []:
            
            if len(property_search_respond)==1: # Unique address            
                list_address_code.append([property_search_respond[0]['code']]) # list 里 第一个 element 是 dict 的key 是 code

            else: # multiple addresses # ie. 2 cooper avenue glen waverley -- gives a list of Dictionary                
                # Initialization: 
                condition=0 # maybe more than 2 states ie 1 0 0, or 0 1 0, or 0 0 1
                mathch=[]
                notmatch=[]

                for each_dict in property_search_respond:                 
                    a=each_dict['code'].replace("-","").upper() #如果是address则含有 / 和 space, code 只有“-”
                    b=df_target.iloc[ind]["address"].replace("+","").upper()
                    c = ("unit"+ b).upper()

                    if "unit" in each_dict["code"] and c[:9]==a[:9]: # 前5+4个character一致 含"unit"
                        mathch.append(each_dict) # 暂存
                        condition +=1 # 已经找到, 更新状态
                        
                    elif "unit" not in each_dict["code"] and a[:5]==b[:5]: # 前5个character一致
                        mathch.append(each_dict) # 暂存
                        condition +=1 # 已经找到, 更新状态
                        
                    else:
                        notmatch.append(each_dict) # 暂存

                if condition>0: # 等到 for lopp 完成之后
                    list_address_code.append([mathch[0]['code']])
                    notmatch=[] # Reset 暂存 # 多个中 找到一个完全match的 那么其他近似的都不要了

                else: # 完全 match 不上的，存下来
                    list_address_code.append([notmatch]) # 把dict 集中在一个 dataframe 的 cell里，list of dicts
                    # 非常注意：能不是 notmatch[0] -- 不然只选了这个list里面第一个 element (一个dict)
        else:
            list_address_code.append([])


# At the end, convert the list to dataframe:
df_temp_addresscode = pd.DataFrame (list_address_code, columns = ['address_code'])    

# update the master dataframe:
updated_df=pd.concat([filesaved_df,df_temp_addresscode])
updated_df_resetindex = updated_df.reset_index(drop=True)

# create excel writer object
writer=pd.ExcelWriter(f'Address_Code_Output_{suburb}.xlsx')
# write dataframe to excel
updated_df_resetindex.to_excel(writer, index=False)
# save the excel
writer.save()
print('DataFrame is written successfully to Excel File.')

print("-------------------------------------------")
print(updated_df_resetindex.info())
print("-------------------------------------------")
print(f"we have {len(list_address_code)} recorded in list_address_code") # 等于 API call 次数，因为空[]也放进去了
print(f"the API Used in this run is: {api_count}")
print(f"lenght of master df is now: {len(updated_df_resetindex)}")
print("-------------------------------------------")
print(updated_df_resetindex.tail(2))
print("-------------------------------------------")
# print(notmatch)

DataFrame is written successfully to Excel File.
-------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1395 entries, 0 to 1394
Data columns (total 1 columns):
address_code    1395 non-null object
dtypes: object(1)
memory usage: 11.0+ KB
None
-------------------------------------------
we have 50 recorded in list_address_code
the API Used in this run is: 50
lenght of master df is now: 1395
-------------------------------------------
                               address_code
1393    56-medina-rd-glen-waverley-vic-3150
1394  6-merrick-cres-glen-waverley-vic-3150
-------------------------------------------


In [14]:
# updated_df_resetindex.tail(10)
# str(['2-cooper-ave-glen-waverley-vic-3150'][0])
# len("2COOPERAVEGLENWAVERLEYVIC3150")
# len("2COOPERAVENUEGLENWAVERLEY3150")

In [15]:
'''
之前无法准确找到，因为有部分 respond给出 list of dict 多个地址：

a=property_search_respond[0]['code'].replace("-","").upper()

b=df_target.iloc[3461]["address"].replace("+","").upper()


# 前是个 character match
a[:8]==b[:8]

'''

'\n之前无法准确找到，因为有部分 respond给出 list of dict 多个地址：\n\na=property_search_respond[0][\'code\'].replace("-","").upper()\n\nb=df_target.iloc[3461]["address"].replace("+","").upper()\n\n\n# 前是个 character match\na[:8]==b[:8]\n\n'

In [16]:
'''
test=[{"address":"Cherry","code":"5-cherry"},{"address":"xxxxxxx","code":"fasfafsaf"}]
for each in test:
    print(each["code"])
    print(type(each["code"]))
'''

'\ntest=[{"address":"Cherry","code":"5-cherry"},{"address":"xxxxxxx","code":"fasfafsaf"}]\nfor each in test:\n    print(each["code"])\n    print(type(each["code"]))\n'

# Check: address code

非常注意：len 涵盖了 0 row, 而 iloc 不含，那么如果 len = 18, iloc最大范围是17 ！！！！

In [17]:
# 非常注意：len 涵盖了 0 row, 而 iloc 不含，那么如果 len = 18, iloc最大范围是17
check_content = updated_df.iloc[len(updated_df_resetindex)-1]["address_code"]
check_content

'6-merrick-cres-glen-waverley-vic-3150'

In [18]:
df_target.iloc[len(updated_df_resetindex)-1]["address"]

'6+MERRICK+Cres+GLEN+WAVERLEY+3150'

# Concat address and address code:

In [19]:
# df_target.iloc[0:len(updated_df_resetindex)-1]

In [20]:
df_combined_addresscode = pd.concat([df_target.iloc[0:len(updated_df_resetindex)],updated_df_resetindex.iloc[0:len(updated_df_resetindex)]],sort=False,axis=1)
df_combined_addresscode.tail(6)

address  \
1389  2+MATTHEW+STREET+GLEN+WAVERLEY+3150   
1390       11+THE+RISE+GLEN+WAVERLEY+3150   
1391       14+THE+RISE+GLEN+WAVERLEY+3150   
1392   3+GWINGANA+Cres+GLEN+WAVERLEY+3150   
1393    56+MEDINA+ROAD+GLEN+WAVERLEY+3150   
1394    6+MERRICK+Cres+GLEN+WAVERLEY+3150   

                                address_code  
1389     2-matthew-st-glen-waverley-vic-3150  
1390      11-the-rise-glen-waverley-vic-3150  
1391      14-the-rise-glen-waverley-vic-3150  
1392  3-gwingana-cres-glen-waverley-vic-3150  
1393     56-medina-rd-glen-waverley-vic-3150  
1394   6-merrick-cres-glen-waverley-vic-3150

#### 学习：非常注意

In [21]:
# 这个找的是 row number, 含 header，所以上面 concat中，不需要减去1，含header已经少了一个！！！！！
df_target.iloc[0:0]

Empty DataFrame
Columns: [address]
Index: []

In [22]:
# 这个找的是 index
df_target.iloc[0]["address"]

'1+22+MELALEUCA+DRIVE+GLEN+WAVERLEY+3150'

#### Missing Entry from API:

In [23]:
(df_combined_addresscode["address_code"]=="[]").sum()

2

In [24]:
df_combined_addresscode[df_combined_addresscode["address_code"]=="[]"]

address address_code
90      2+5+MIMOSA+STREET+GLEN+WAVERLEY+3150           []
304  1+13+PANORAMIC+GROVE+GLEN+WAVERLEY+3150           []

# Save Combined Address to Excel:

In [25]:
# create excel writer object
writer=pd.ExcelWriter(f'Address_Code_Output_Combined_{suburb}.xlsx')
# write dataframe to excel
df_combined_addresscode.to_excel(writer, index=False)
# save the excel
writer.save()
print('DataFrame is written successfully to Excel File.')

DataFrame is written successfully to Excel File.
